File: src/antiplagiat.ts
/**
 * Оценка по фрагментам в духе модуля поиска сгенерированного текста
 * системы «Антиплагиат».
 *
 * Настоящий классификатор закрыт, поэтому здесь приближение: каждый
 * фрагмент (абзац или склейка коротких абзацев) описывается набором
 * интерпретируемых признаков, логистическая модель переводит их в
 * вероятность, а доля ИИ-текста считается как доля знаков во фрагментах
 * выше порога — так же выглядит итог в отчёте системы.
 *
 * Веса по умолчанию подобраны вручную. Команда `calibrate` дообучает их
 * на фрагментах, которые система реально подсветила в ваших отчётах.
 */

import { analyzeInternal, WEIGHTS } from "./detect.ts";
import { cv, lineCol, mattr, mean, sentences, words } from "./text.ts";
import type { AnalyzeOptions, Issue } from "./types.ts";

export const FEATURE_NAMES = [
  "плотность примет",
  "однообразие длины предложений",
  "типичная длина предложения",
  "канцелярит",
  "бедная пунктуация",
  "однообразные начала предложений",
  "бедный словарь",
] as const;

export interface Model {
  bias: number;
  weights: number[];
  threshold: number;
}

export const DEFAULT_MODEL: Model = {
  bias: -4.4,
  weights: [1.1, 2.6, 1.4, 0.35, 1.2, 1.3, 1.6],
  threshold: 0.5,
};

export interface Fragment {
  start: number;
  end: number;
  line: number;
  endLine: number;
  words: number;
  features: number[];
  probability: number;
  ai: boolean;
  reasons: string[];
  preview: string;
}

export interface AntiplagiatReport {
  /** Доля знаков текста во фрагментах, помеченных как ИИ, 0–100. */
  aiShare: number;
  fragments: Fragment[];
  suspicious: boolean;
  suspiciousReasons: string[];
  model: "default" | "calibrated";
  threshold: number;
}

const MIN_FRAGMENT_WORDS = 40;

function sigmoid(z: number): number {
  return 1 / (1 + Math.exp(-z));
}

export function predict(model: Model, f: number[]): number {
  let z = model.bias;
  for (let i = 0; i < f.length; i += 1) z += (model.weights[i] ?? 0) * (f[i] ?? 0);
  return sigmoid(z);
}

export function featuresFor(text: string, fragmentIssues: Issue[]): number[] {
  const ws = words(text);
  const n = Math.max(ws.length, 1);
  const ss = sentences(text).filter((s) => s.words >= 2);
  const lens = ss.map((s) => s.words);

  const signal = fragmentIssues.filter((i) => !i.styleOnly).reduce((a, i) => a + (WEIGHTS[i.type] ?? 1), 0);
  const density = Math.min(4, (signal * 100) / n / 3);
  const uniformity = lens.length >= 3 ? Math.max(0, 1 - cv(lens) / 0.6) : 0.3;
  const m = mean(lens);
  const typical = lens.length ? Math.exp(-(((m - 19) / 7) ** 2)) : 0;
  const clerical = Math.min(3, (fragmentIssues.filter((i) => i.styleOnly).length * 100) / n / 2);
  const rare = ss.filter((s) => /[;:()!?…]|\s—\s/.test(s.text)).length;
  const poorPunct = ss.length >= 3 ? Math.max(0, 1 - rare / ss.length / 0.5) : 0.3;
  const firsts = ss.map((s) => (words(s.text)[0] ?? "").toLowerCase());
  const openerRepeat = firsts.length >= 3 ? 1 - new Set(firsts).size / firsts.length : 0;
  const diversity = ws.length >= 30 ? mattr(ws, Math.min(50, ws.length)) : 0.8;
  const poorVocab = Math.max(0, Math.min(1, (0.85 - diversity) / 0.2));
  return [density, uniformity, typical, clerical, poorPunct, openerRepeat, poorVocab];
}

function reasonsFor(model: Model, f: number[]): string[] {
  return f
    .map((v, i) => ({ i, c: v * (model.weights[i] ?? 0) }))
    .filter((x) => x.c > 0.5)
    .sort((a, b) => b.c - a.c)
    .slice(0, 3)
    .map((x) => FEATURE_NAMES[x.i] ?? "");
}

interface RawFragment {
  start: number;
  end: number;
  text: string;
}

/** Абзацы прозы и списков; короткие склеиваются со следующими. */
function fragmentsOf(internals: ReturnType<typeof analyzeInternal>["internals"]): RawFragment[] {
  const out: RawFragment[] = [];
  let cur: RawFragment | null = null;
  for (const b of internals.blocks) {
    if (b.kind !== "prose" && b.kind !== "list") {
      if (b.kind === "heading" && cur && words(cur.text).length >= MIN_FRAGMENT_WORDS / 2) {
        out.push(cur);
        cur = null;
      }
      continue;
    }
    if (!cur) cur = { start: b.start, end: b.end, text: b.text };
    else {
      cur.end = b.end;
      cur.text += `\n\n${b.text}`;
    }
    if (words(cur.text).length >= MIN_FRAGMENT_WORDS) {
      out.push(cur);
      cur = null;
    }
  }
  if (cur) {
    const last = out[out.length - 1];
    if (last && words(cur.text).length < MIN_FRAGMENT_WORDS / 2) {
      last.end = cur.end;
      last.text += `\n\n${cur.text}`;
    } else out.push(cur);
  }
  return out;
}

export function antiplagiat(source: string, options: AnalyzeOptions & { model?: Model } = {}): AntiplagiatReport {
  const { result, internals } = analyzeInternal(source, { context: options.context ?? "academic" });
  const model = options.model ?? DEFAULT_MODEL;
  const p = internals.prepared;
  const raws = fragmentsOf(internals);
  const toSrc = (i: number): number => p.toSource[i] ?? i;
  let aiChars = 0;
  let allChars = 0;
  const fragments: Fragment[] = raws.map((r) => {
    const s = toSrc(r.start);
    const e = toSrc(r.end);
    const inside = result.issues.filter((i) => i.index >= s && i.index < e);
    const f = featuresFor(r.text, inside);
    const prob = predict(model, f);
    const chars = r.text.replace(/\s+/g, " ").trim().length;
    allChars += chars;
    const ai = prob >= model.threshold;
    if (ai) aiChars += chars;
    const lineOf = (idx: number): number => lineCol(p.lineStarts, idx).line;
    return {
      start: s,
      end: e,
      line: lineOf(s),
      endLine: lineOf(e),
      words: words(r.text).length,
      features: f.map((v) => Number(v.toFixed(3))),
      probability: Number(prob.toFixed(3)),
      ai,
      reasons: reasonsFor(model, f),
      preview: p.source.slice(s, Math.min(e, s + 90)).replace(/\s+/g, " ").trim(),
    };
  });

  const suspiciousReasons: string[] = [];
  if (p.invisible.length) suspiciousReasons.push(`невидимые символы: ${p.invisible.length}`);
  if (p.homoglyphs.length) suspiciousReasons.push(`слова со смешанной латиницей и кириллицей: ${p.homoglyphs.length}`);

  return {
    aiShare: allChars ? Number(((aiChars * 100) / allChars).toFixed(1)) : 0,
    fragments,
    suspicious: suspiciousReasons.length > 0,
    suspiciousReasons,
    model: options.model ? "calibrated" : "default",
    threshold: model.threshold,
  };
}

// ─── Калибровка по реальным отчётам ─────────────────────────────────────

export interface Sample {
  f: number[];
  y: 0 | 1;
}

export interface CalibrationFile {
  version: 1;
  model: Model;
  samples: Sample[];
}

function norm(s: string): string {
  return s
    .toLowerCase()
    .replace(/ё/g, "е")
    .replace(/[*_`#>|]/g, "")
    .replace(/[«»„“”"]/g, "")
    .replace(/\s+/g, " ")
    .trim();
}

/**
 * Размечает фрагменты документа по кускам, которые система подсветила
 * в отчёте (скопированы в текстовый файл, разделены пустой строкой).
 */
export function labelFragments(source: string, marked: string[]): Sample[] {
  const { internals, result } = analyzeInternal(source, { context: "academic" });
  const p = internals.prepared;
  const keys = marked.map(norm).filter((m) => m.length >= 30).map((m) => m.slice(0, 60));
  return fragmentsOf(internals).map((r) => {
    const s = p.toSource[r.start] ?? r.start;
    const e = p.toSource[r.end] ?? r.end;
    const inside = result.issues.filter((i) => i.index >= s && i.index < e);
    const text = norm(p.source.slice(s, e));
    const y: 0 | 1 = keys.some((k) => text.includes(k)) ? 1 : 0;
    return { f: featuresFor(r.text, inside), y };
  });
}

/** Логистическая регрессия с L2 и выбором порога по сбалансированной точности. */
export function fit(samples: Sample[], start: Model = DEFAULT_MODEL): Model {
  const pos = samples.filter((s) => s.y === 1).length;
  const neg = samples.length - pos;
  if (pos === 0 || neg === 0) return start;
  const wPos = samples.length / (2 * pos);
  const wNeg = samples.length / (2 * neg);
  const w = [...start.weights];
  let b = start.bias;
  const lr = 0.05;
  const l2 = 0.01;
  for (let it = 0; it < 4000; it += 1) {
    const gw = new Array<number>(w.length).fill(0);
    let gb = 0;
    for (const s of samples) {
      const pr = predict({ bias: b, weights: w, threshold: 0.5 }, s.f);
      const err = (pr - s.y) * (s.y === 1 ? wPos : wNeg);
      for (let k = 0; k < w.length; k += 1) gw[k] = (gw[k] ?? 0) + err * (s.f[k] ?? 0);
      gb += err;
    }
    for (let k = 0; k < w.length; k += 1) {
      w[k] = (w[k] ?? 0) - lr * ((gw[k] ?? 0) / samples.length + l2 * ((w[k] ?? 0) - (start.weights[k] ?? 0)));
    }
    b -= (lr * gb) / samples.length;
  }
  const model: Model = { bias: b, weights: w, threshold: 0.5 };
  let best = { t: 0.5, score: -1 };
  for (let t = 0.2; t <= 0.8001; t += 0.02) {
    let tp = 0;
    let tn = 0;
    for (const s of samples) {
      const hit = predict(model, s.f) >= t;
      if (hit && s.y === 1) tp += 1;
      if (!hit && s.y === 0) tn += 1;
    }
    const bal = (tp / pos + tn / neg) / 2;
    if (bal > best.score) best = { t, score: bal };
  }
  model.threshold = Number(best.t.toFixed(2));
  model.bias = Number(model.bias.toFixed(4));
  model.weights = model.weights.map((x) => Number(x.toFixed(4)));
  return model;
}

export function balancedAccuracy(model: Model, samples: Sample[]): number {
  const pos = samples.filter((s) => s.y === 1);
  const neg = samples.filter((s) => s.y === 0);
  if (!pos.length || !neg.length) return Number.NaN;
  const tp = pos.filter((s) => predict(model, s.f) >= model.threshold).length;
  const tn = neg.filter((s) => predict(model, s.f) < model.threshold).length;
  return (tp / pos.length + tn / neg.length) / 2;
}

File: src/aiw_ru/antiplagiat.py
"""Оценка по фрагментам в духе модуля поиска сгенерированного текста системы «Антиплагиат».

Настоящий классификатор закрыт, поэтому здесь приближение: каждый фрагмент
(абзац или склейка коротких абзацев) описывается набором интерпретируемых
признаков, логистическая модель переводит их в вероятность, а доля ИИ-текста
считается как доля знаков во фрагментах выше порога — так же выглядит итог в
отчёте системы.

Веса по умолчанию подобраны вручную. Команда `calibrate` дообучает их на
фрагментах, которые система реально подсветила в ваших отчётах.
"""

from __future__ import annotations

import math
from dataclasses import dataclass
from typing import Literal

from pydantic import Field

from aiw_ru.compat import fixed, jsre, total, trim
from aiw_ru.detect import WEIGHTS, Internals, analyze_internal
from aiw_ru.text import cv, line_col, mattr, mean, sentences, words
from aiw_ru.types import ContextMode, Issue, Record

FEATURE_NAMES: tuple[str, ...] = (
    "плотность примет",
    "однообразие длины предложений",
    "типичная длина предложения",
    "канцелярит",
    "бедная пунктуация",
    "однообразные начала предложений",
    "бедный словарь",
)


class Model(Record):
    bias: float
    weights: list[float]
    threshold: float


DEFAULT_MODEL = Model(bias=-4.4, weights=[1.1, 2.6, 1.4, 0.35, 1.2, 1.3, 1.6], threshold=0.5)


class Fragment(Record):
    start: int
    end: int
    line: int
    end_line: int
    words: int
    features: list[float]
    probability: float
    ai: bool
    reasons: list[str]
    preview: str


class AntiplagiatReport(Record):
    # Доля знаков текста во фрагментах, помеченных как ИИ, 0–100.
    ai_share: float
    fragments: list[Fragment]
    suspicious: bool
    suspicious_reasons: list[str]
    model: Literal["default", "calibrated"]
    threshold: float


MIN_FRAGMENT_WORDS = 40
RARE_PUNCT_RE = jsre("[;:()!?…]|\\s—\\s")
SPACES_RE = jsre("\\s+", "g")


def _sigmoid(z: float) -> float:
    return 1 / (1 + math.exp(-z))


def predict(model: Model, f: list[float]) -> float:
    z = model.bias
    for i, v in enumerate(f):
        z += (model.weights[i] if i < len(model.weights) else 0) * v
    return _sigmoid(z)


def features_for(text: str, fragment_issues: list[Issue]) -> list[float]:
    ws = words(text)
    n = max(len(ws), 1)
    ss = [s for s in sentences(text) if s.words >= 2]
    lens = [s.words for s in ss]

    signal = total(WEIGHTS.get(i.type, 1) for i in fragment_issues if not i.style_only)
    density = min(4, signal * 100 / n / 3)
    uniformity = max(0, 1 - cv(lens) / 0.6) if len(lens) >= 3 else 0.3
    m = mean(lens)
    typical = math.exp(-(((m - 19) / 7) ** 2)) if lens else 0
    clerical = min(3, sum(1 for i in fragment_issues if i.style_only) * 100 / n / 2)
    rare = sum(1 for s in ss if RARE_PUNCT_RE.search(s.text))
    poor_punct = max(0, 1 - rare / len(ss) / 0.5) if len(ss) >= 3 else 0.3
    firsts = [(words(s.text) or [""])[0].lower() for s in ss]
    opener_repeat = 1 - len(set(firsts)) / len(firsts) if len(firsts) >= 3 else 0
    diversity = mattr(ws, min(50, len(ws))) if len(ws) >= 30 else 0.8
    poor_vocab = max(0, min(1, (0.85 - diversity) / 0.2))
    return [density, uniformity, typical, clerical, poor_punct, opener_repeat, poor_vocab]


def _reasons_for(model: Model, f: list[float]) -> list[str]:
    contrib = [(i, v * (model.weights[i] if i < len(model.weights) else 0)) for i, v in enumerate(f)]
    top = sorted((x for x in contrib if x[1] > 0.5), key=lambda x: -x[1])[:3]
    return [FEATURE_NAMES[i] if i < len(FEATURE_NAMES) else "" for i, _ in top]


@dataclass(slots=True)
class _RawFragment:
    start: int
    end: int
    text: str


def _fragments_of(internals: Internals) -> list[_RawFragment]:
    """Абзацы прозы и списков; короткие склеиваются со следующими."""
    out: list[_RawFragment] = []
    cur: _RawFragment | None = None
    for b in internals.blocks:
        if b.kind not in ("prose", "list"):
            if b.kind == "heading" and cur and len(words(cur.text)) >= MIN_FRAGMENT_WORDS / 2:
                out.append(cur)
                cur = None
            continue
        if cur is None:
            cur = _RawFragment(b.start, b.end, b.text)
        else:
            cur.end = b.end
            cur.text += f"\n\n{b.text}"
        if len(words(cur.text)) >= MIN_FRAGMENT_WORDS:
            out.append(cur)
            cur = None
    if cur is not None:
        if out and len(words(cur.text)) < MIN_FRAGMENT_WORDS / 2:
            out[-1].end = cur.end
            out[-1].text += f"\n\n{cur.text}"
        else:
            out.append(cur)
    return out


def _chars(s: str) -> int:
    """Длина текста без лишних пробелов."""
    return len(trim(SPACES_RE.sub(" ", s)))


def antiplagiat(source: str, context: ContextMode = "academic", model: Model | None = None) -> AntiplagiatReport:
    result, internals = analyze_internal(source, context)
    m = model or DEFAULT_MODEL
    p = internals.prepared

    def to_src(i: int) -> int:
        return p.to_source[i] if i < len(p.to_source) else i

    ai_chars = 0
    all_chars = 0
    fragments: list[Fragment] = []
    for r in _fragments_of(internals):
        s, e = to_src(r.start), to_src(r.end)
        inside = [i for i in result.issues if s <= i.index < e]
        f = features_for(r.text, inside)
        prob = predict(m, f)
        chars = _chars(r.text)
        all_chars += chars
        ai = prob >= m.threshold
        if ai:
            ai_chars += chars
        fragments.append(
            Fragment(
                start=s,
                end=e,
                line=line_col(p.line_starts, s)[0],
                end_line=line_col(p.line_starts, e)[0],
                words=len(words(r.text)),
                features=[fixed(v, 3) for v in f],
                probability=fixed(prob, 3),
                ai=ai,
                reasons=_reasons_for(m, f),
                preview=trim(SPACES_RE.sub(" ", p.source[s : min(e, s + 90)])),
            )
        )

    reasons: list[str] = []
    if p.invisible:
        reasons.append(f"невидимые символы: {len(p.invisible)}")
    if p.homoglyphs:
        reasons.append(f"слова со смешанной латиницей и кириллицей: {len(p.homoglyphs)}")
    return AntiplagiatReport(
        ai_share=fixed(ai_chars * 100 / all_chars, 1) if all_chars else 0,
        fragments=fragments,
        suspicious=bool(reasons),
        suspicious_reasons=reasons,
        model="calibrated" if model else "default",
        threshold=m.threshold,
    )


# ─── Калибровка по реальным отчётам ─────────────────────────────────────


class Sample(Record):
    f: list[float]
    y: Literal[0, 1]


class DocFragment(Record):
    f: list[float]
    chars: int


class DocSample(Record):
    """Документ, для которого известна только итоговая доля ИИ-текста из отчёта.

    Разметки фрагментов нет; для каждого фрагмента хранятся признаки и длина в знаках.
    """

    fragments: list[DocFragment]
    # Доля ИИ-текста по отчёту, 0–100.
    share: float


class CalibrationFile(Record):
    version: Literal[1]
    model: Model
    samples: list[Sample] = Field(default_factory=list)
    # Документы с одной итоговой долей (`calibrate --share`).
    documents: list[DocSample] = Field(default_factory=list)


NORM_MARKUP_RE = jsre("[*_`#>|]", "g")
NORM_QUOTES_RE = jsre('[«»„“”"]', "g")


def _norm(s: str) -> str:
    s = s.lower().replace("ё", "е")
    s = NORM_QUOTES_RE.sub("", NORM_MARKUP_RE.sub("", s))
    return trim(SPACES_RE.sub(" ", s))


def label_fragments(source: str, marked: list[str]) -> list[Sample]:
    """Размечает фрагменты документа по кускам, которые система подсветила в отчёте.

    Куски скопированы из отчёта в текстовый файл и разделены пустой строкой.
    """
    result, internals = analyze_internal(source, "academic")
    p = internals.prepared
    keys = [k[:60] for k in (_norm(m) for m in marked) if len(k) >= 30]
    out: list[Sample] = []
    for r in _fragments_of(internals):
        s = p.to_source[r.start] if r.start < len(p.to_source) else r.start
        e = p.to_source[r.end] if r.end < len(p.to_source) else r.end
        inside = [i for i in result.issues if s <= i.index < e]
        text = _norm(p.source[s:e])
        y: Literal[0, 1] = 1 if any(k in text for k in keys) else 0
        out.append(Sample(f=features_for(r.text, inside), y=y))
    return out


def fit(samples: list[Sample], start: Model = DEFAULT_MODEL) -> Model:
    """Логистическая регрессия с L2 и выбором порога по сбалансированной точности."""
    pos = sum(1 for s in samples if s.y == 1)
    neg = len(samples) - pos
    if pos == 0 or neg == 0:
        return start
    w_pos = len(samples) / (2 * pos)
    w_neg = len(samples) / (2 * neg)
    w = list(start.weights)
    b = start.bias
    lr = 0.05
    l2 = 0.01
    for _ in range(4000):
        gw = [0.0] * len(w)
        gb = 0.0
        cur = Model(bias=b, weights=w, threshold=0.5)
        for s in samples:
            err = (predict(cur, s.f) - s.y) * (w_pos if s.y == 1 else w_neg)
            for k in range(len(w)):
                gw[k] += err * (s.f[k] if k < len(s.f) else 0)
            gb += err
        for k in range(len(w)):
            w[k] = w[k] - lr * (gw[k] / len(samples) + l2 * (w[k] - start.weights[k]))
        b -= lr * gb / len(samples)
    model = Model(bias=b, weights=w, threshold=0.5)
    best_t, best_score = 0.5, -1.0
    t = 0.2
    while t <= 0.8001:
        tp = tn = 0
        for s in samples:
            hit = predict(model, s.f) >= t
            if hit and s.y == 1:
                tp += 1
            if not hit and s.y == 0:
                tn += 1
        bal = (tp / pos + tn / neg) / 2
        if bal > best_score:
            best_t, best_score = t, bal
        t += 0.02
    return Model(bias=fixed(model.bias, 4), weights=[fixed(x, 4) for x in model.weights], threshold=fixed(best_t, 2))


def balanced_accuracy(model: Model, samples: list[Sample]) -> float:
    pos = [s for s in samples if s.y == 1]
    neg = [s for s in samples if s.y == 0]
    if not pos or not neg:
        return math.nan
    tp = sum(1 for s in pos if predict(model, s.f) >= model.threshold)
    tn = sum(1 for s in neg if predict(model, s.f) < model.threshold)
    return (tp / len(pos) + tn / len(neg)) / 2


def document_sample(source: str, share: float) -> DocSample:
    """Признаки и длины фрагментов документа с долей ИИ-текста из отчёта."""
    report = antiplagiat(source)
    return DocSample(
        fragments=[DocFragment(f=f.features, chars=_chars(source[f.start : f.end])) for f in report.fragments],
        share=share,
    )


def predict_share(model: Model, doc: DocSample) -> float:
    """Доля ИИ-текста по знакам, которую модель предскажет для документа."""
    ai = 0
    all_ = 0
    for fr in doc.fragments:
        all_ += fr.chars
        if predict(model, fr.f) >= model.threshold:
            ai += fr.chars
    return ai * 100 / all_ if all_ else 0


def share_error(model: Model, docs: list[DocSample]) -> float:
    """Средняя ошибка доли ИИ-текста в процентных пунктах."""
    if not docs:
        return math.nan
    return total(abs(predict_share(model, d) - d.share) for d in docs) / len(docs)


def fit_share(model: Model, docs: list[DocSample]) -> Model:
    """Подстройка под итоговые доли из отчётов.

    Сдвиг свободного члена, при котором предсказанные доли ближе всего к отчётным.
    Веса признаков не меняются: одной цифры на документ хватает только на общий
    уровень строгости системы.
    """
    if not docs:
        return model
    best_bias, best_err = model.bias, share_error(model, docs)
    d = -8.0
    while d <= 8.0001:
        bias = model.bias + d
        err = share_error(model.model_copy(update={"bias": bias}), docs)
        closer = abs(d) < abs(best_bias - model.bias)
        if err < best_err - 1e-9 or (abs(err - best_err) <= 1e-9 and closer):
            best_bias, best_err = bias, err
        d += 0.02
    return model.model_copy(update={"bias": fixed(best_bias, 4)})
